presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyStateFixedHistogramJacknifeStrategy.java

```diff
@@ -0,0 +1,155 @@
+
+import java.util.HashMap;
+import java.util.Map;
+
+/*
+Calculates sample entropy using jacknife estimates on a fixed histogram.
+See http://cs.brown.edu/~pvaliant/unseen_nips.pdf.
+ */
+public class DifferentialEntropyStateFixedHistogramJacknifeStrategy
+        extends DifferentialEntropyStateFixedHistogramStrategy
+{
+    public DifferentialEntropyStateFixedHistogramJacknifeStrategy(long bucketCount, double min, double max)
+    {
+        super(new FixedDoubleBreakdownHistogram((int) bucketCount, min, max));
+    }
+
+    protected DifferentialEntropyStateFixedHistogramJacknifeStrategy(DifferentialEntropyStateFixedHistogramJacknifeStrategy other)
+    {
+        super(new FixedDoubleBreakdownHistogram(other.getBreakdownHistogram()));
+    }
+
+    public DifferentialEntropyStateFixedHistogramJacknifeStrategy(SliceInput input)
+    {
+        super(new FixedDoubleBreakdownHistogram(input));
+    }
+
+    @Override
+    public void validateParams(
+            long bucketCount,
+            double sample,
+            Double weight,
+            Double min,
+            Double max)
+    {
+        super.validateParams(bucketCount, sample, weight, min, max);
+    }
+
+    @Override
+    public void mergeWith(DifferentialEntropyStateStrategy other)
+    {
+        getBreakdownHistogram()
+                .mergeWith(((DifferentialEntropyStateFixedHistogramJacknifeStrategy) other).getBreakdownHistogram());
+    }
+
+    @Override
+    public void add(double value, double weight)
+    {
+        getBreakdownHistogram().add(value, weight);
+    }
+
+    @Override
+    public double calculateEntropy()
+    {
+        final Map<Double, Double> bucketWeights = new HashMap<Double, Double>();
+        Streams.stream(getBreakdownHistogram().iterator()).forEach(
+                e -> {
+                    bucketWeights.put(
+                            e.left,
+                            e.breakdown.entrySet().stream().mapToDouble(w -> w.getKey() * w.getValue()).sum());
+                });
+        final double sumW =
+                bucketWeights.values().stream().mapToDouble(w -> w).sum();
+        if (sumW == 0.0) {
+            return 0.0;
+        }
+        final long n = Streams.stream(getBreakdownHistogram())
+                .mapToLong(e -> e.breakdown.values().stream().mapToLong(i -> i).sum())
+                .sum();
+        final double sumWLogW =
+                bucketWeights.values().stream().mapToDouble(w -> w == 0.0 ? 0.0 : w * Math.log(w)).sum();
+
+        double entropy = n * calculateEntropy(histogram.getWidth(), sumW, sumWLogW);
+        entropy -= Streams.stream(getBreakdownHistogram().iterator()).mapToDouble(
+                e -> {
+                    final double bucketWeight = bucketWeights.get(e.left);
+                    if (bucketWeight == 0.0) {
+                        return 0.0;
+                    }
+                    return e.breakdown.entrySet().stream().mapToDouble(
+                            bucketE -> {
```

Comment:
I think it would be more readable if we break this into a static method and use a method reference here.  Also, the variable names should be cleaned up to something more descriptive.